src/py/mat3ra/made/tools/analyze.py

```diff
@@ -324,3 +337,30 @@ def get_atomic_coordinates_extremum(
     coordinates = new_material.basis.coordinates.to_array_of_values_with_ids()
     values = [coord.value[{"x": 0, "y": 1, "z": 2}[axis]] for coord in coordinates]
     return getattr(np, extremum)(values)
+
+
+def get_undercoordinated_atoms(
+    material: Material, indices_to_check: Optional[List[int]] = None, coordination_threshold: Optional[int] = None
+) -> Tuple[List[int], dict]:
+    if indices_to_check is None:
+        indices_to_check = material.basis.coordinates.ids
+
+    coordinates_array = material.basis.coordinates.values
+    number_of_neighbors = []
+    atom_neighbors_info = {}
+    for idx in indices_to_check:
+        coordinate = coordinates_array[idx]
+        neighbors_indices = get_nearest_neighbors_atom_indices(material=material, coordinate=coordinate, cutoff=6) or []
+        neighbors_coordinates = [coordinates_array[i] for i in neighbors_indices]
+        neighbors_center = get_center_of_coordinates(neighbors_coordinates)
+        atom_neighbors_info[idx] = (len(neighbors_indices), neighbors_indices, neighbors_center)
```

Comment:
This needs a data structure and maybe a class - `AtomicSiteEnvironment` or something like that